### PR TITLE
Distributes partition-to-be table before ProcessUtility

### LINKED ISSUE #4483: Outer join error when creating partitions for a distributed table with a foreign key to a co-located table
We do not support the outer join that happens to validate the foreign key when attaching a partition to a distributed-partitioned table with a foreign key to another distributed table:
``` sql
drop table if exists events, test;
create table test (x int primary key, y int);
select create_distributed_table('test','x');

create table events (event_id serial not null references test (x), event_time timestamptz not null default now(), a int, payload jsonb) partition by range (event_time);
select create_distributed_table('events','event_id');

select partman.create_parent('public.events', 'event_time', 'native', '1 hour');

ERROR:  cannot pushdown the subquery
CONTEXT: SQL statement "SELECT fk."event_id" FROM ONLY "public"."events_p2021_01_08_0400" fk LEFT OUTER JOIN ONLY "public"."test" pk ON ( pk."x" OPERATOR(pg_catalog.=) fk."event_id") WHERE pk."x" IS NULL AND (fk."event_id" IS NOT NULL)"
SQL statement "ALTER TABLE public.events ATTACH PARTITION public.events_p2021_01_08_0400 FOR VALUES FROM ('2021-01-08 04:00:00+01') TO ('2021-01-08 05:00:00+01')"
PL/pgSQL function create_partition_time(text,timestamp with time zone[],boolean,boolean) line 243 at EXECUTE
PL/pgSQL function partman.create_parent(text,text,text,text,text[],integer,text,text,boolean,text,text,text[],boolean,text,boolean,boolean) line 472 at assignment
DETAIL: Complex subqueries and CTEs cannot be in the outer part of the outer join
HINT: 
CONTEXT: PL/pgSQL function create_partition_time(text,timestamp with time zone[],boolean,boolean) line 482 at RAISE
PL/pgSQL function partman.create_parent(text,text,text,text,text[],integer,text,text,boolean,text,text,text[],boolean,text,boolean,boolean) line 472 at assignment
DETAIL: 
HINT: 
CONTEXT:  PL/pgSQL function partman.create_parent(text,text,text,text,text[],integer,text,text,boolean,text,text,text[],boolean,text,boolean,boolean) line 770 at RAISE
```

### PR DESCRIPTION
**Pull request by naisila**
DESCRIPTION: Fixes a bug that prevents attaching partitions when colocated foreign key exists

The error in the issue #4483 occurs during fk validation of the partition because the partition-to-be table is not distributed.

### Current approach:
If in an `ALTER TABLE .. ATTACH PARTITION` command, the partition-to-be is not distributed, distribute the partition-to-be table before going into `standard_ProcessUtility`, therefore before the fk validation. Details:

- We were distributing the partition in `PostprocessAlterTableStmtAttachPartition` logic: we move that logic to a preprocess function that runs even before `PreprocessAlterTableStmt` -> `PreprocessAlterTableStmtAttachPartition`
- From what I have checked, the only affected part by this change are the changes from the merged PR #5131 . ~~For this reason I added the Oid of the parent-to-be table as an argument to CreateDistributedTable, to treat the partition-to-be table that we distribute as if it were a partition already in order to make the necessary checks.~~ For this reason, we make sure that we pass in the correct distribution column definition for the partition-to-be table.

### Previous approach:
Skips ALTER TABLE constraint checks while planning since we ignore them in the executor anyway. Details:

- A "hacky solution" that has a specific exception on top of the planner. The idea came from the current "a bit dirty" status of the executor as well where ALTER TABLE constraint checks are ignored. See merged PR #2629 
- You can check this solution from the first commit
- We can check for SELECT validation queries in ALTER TABLE _earlier_ in the planner. Currently, I have put the check right before it would normally error out.

### So ... 
I would appreciate if the reviewer picks their "favorite" approach - maybe the current solution to current approach can be improved to be cleaner.

Fixes #4483

### CODE REVIEW

File: src/backend/distributed/planner/query_pushdown_planning.c [hunk 1]
```diff
@@ -811,6 +813,18 @@ DeferredErrorIfUnsupportedRecurringTuplesJoin(
...
+				/*
+				 * These constraints will be validated in worker nodes, so no
+				 * need to run them on coordinator. In fact, we disable their
+				 * execution: check CitusExecutorRun. We don't care whether
+				 * we can pushdown the subquery or not. Hence we simply
+				 * consider this as a supported Recurring Tuples Join.
+				 */
+				if (SimpleAlterTableConstraintCheck(plannerInfo->parse))
+				{
+					continue;
```
Comment (marcocitus) [inline src/backend/distributed/planner/query_pushdown_planning.c:825]: I feel a bit uncomfortable with generating an incorrect query plan and still doing some amount of work. Is there a way to skip the work in CitusExecutorRun?

Comment (naisila) [reply src/backend/distributed/planner/query_pushdown_planning.c:825]: This work is already skipped in CitusExecutorRun.

So the error happens because the partition is not distributed but its a local PG table. If we run the command with an already distributed partition, it doesn't error out in planning phase and all the work is done for it => it is considered as proper pushdown query. But then it is skipped in CitusExecutorRun anyway.

Comment (marcocitus) [reply src/backend/distributed/planner/query_pushdown_planning.c:825]: Got it, so we mainly want to avoid the planner throwing an error. Could we generate some dummy plan early on in the planner?

Comment (marcocitus) [reply src/backend/distributed/planner/query_pushdown_planning.c:825]: Or, could we treat it as a not-distributed query?

Comment (naisila) [reply src/backend/distributed/planner/query_pushdown_planning.c:825]: Thanks for your input. If I can figure out a way early on in the planner to not consider it as a distributed query that would make more sense as it aligns with skipping the query in CitusExecutorRun.

**Comment by onderkalaci** [comment]
> DESCRIPTION: Distribute partition-to-be table before ProcessUtility

It seems it'd be hard to follow for users in the release notes. Maybe something like `Fixes a bug that prevents attaching partitions when colocated foreign key exists`

**Comment by onderkalaci** [comment]
@naisila I think it is worth testing the fix with pg_partman if you have not already done

File: src/test/regress/sql/multi_partitioning.sql [hunk 2]
```diff
@@ -1303,7 +1303,23 @@ SELECT worker_partitioned_relation_total_size(oid) FROM pg_class WHERE relname L
...
 DROP TABLE "events.Energy Added";
 
+-- test we skip the foreign key validation query on coordinator
+-- that happens when attaching a non-distributed partition to a distributed-partitioned table
+-- with a foreign key to another distributed table
+SET search_path = partitioning_schema;
+SET citus.shard_replication_factor = 1;
+CREATE TABLE another_distributed_table (x int primary key, y int);
+SELECT create_distributed_table('another_distributed_table','x');
+CREATE TABLE distributed_parent_table (
```
Comment (onderkalaci) [inline src/test/regress/sql/multi_partitioning.sql:1313]: I think a test when a default partition exists might be interesting as well. PG behaves slightly differently when default partition exists.

File: src/backend/distributed/commands/table.c [hunk 3]
```diff
@@ -399,8 +399,8 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
  * ATTACH PARTITION OF command.
  */
 List *
-PostprocessAlterTableStmtAttachPartition(AlterTableStmt *alterTableStatement,
-										 const char *queryString)
+PreprocessAlterTableStmtAttachPartition(AlterTableStmt *alterTableStatement,
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:402]: I think the major drawback of switching from Pre hook to post hook, the errors are going to be detected on the workers instead of on the parents.

For example, in the following example, the error is detected on the shards where it is detected on the coordinator before the patch:
```SQL
CREATE TABLE parent (
  event_id serial NOT NULL,-- REFERENCES another_table (x),
  event_time timestamptz NOT NULL DEFAULT now())
  PARTITION BY RANGE (event_time);
SELECT create_distributed_table('parent', 'event_id');


CREATE TABLE non_distributed_child_very_very_very_loooooooooooooooooooooooooooooooooong_table_name (event_id int NOT NULL, event_time timestamptz NOT NULL DEFAULT now(), extra_column_that_doesnt_exist_on_parent int);


ALTER TABLE parent ATTACH PARTITION non_distributed_child_very_very_very_loooooooooooooooooooooooooooooooooong_table_name FOR VALUES FROM ('2021-06-01') TO ('2021-07-30');
ERROR:  table "non_distributed_child_very_very_very_looooooooooooooooooooooooo" contains column "extra_column_that_doesnt_exist_on_parent" not found in parent "parent"
DETAIL:  The new partition may contain only the columns present in parent.
```

In general, we prefer to do things in the post-hook for this reason. 

Do you see any check in `ATExecAttachPartition` that we'd rather prefer to do on the coordinator?

Similarly, when we do things in the post hook, we ensure that the proper locks are acquired by Postgres on the tables. Now, they are not acquired at this step.  Do you see any issues with that?

**Review by onderkalaci — COMMENTED** [review]
It is annoying that even after this patch, the users could hit some issues when (a) the partition has already data (b) the name of the partition is long   (hence switch to sequential execution mode).

Can you see an easy way to fix that? I guess it is related to https://github.com/citusdata/citus/issues/3895 where `create_distributed_table` internally forces parallel execution. Still, with https://github.com/citusdata/citus/pull/3841, we should **probably** be fine to remove forcing to use parallelization and/or the sequential execution checks.

Anyway, steps to repro the issue for reference:
```SQL
CREATE TABLE another_table (x int primary key, y int);
SELECT create_distributed_table('another_table','x');

INSERT INTO another_table SELECT i,i from generate_series(0,100)i;


CREATE TABLE parent (
  event_id serial NOT NULL REFERENCES another_table (x),
  event_time timestamptz NOT NULL DEFAULT now())
  PARTITION BY RANGE (event_time);
SELECT create_distributed_table('parent', 'event_id');


CREATE TABLE non_distributed_child_very_very_very_loooooooooooooooooooooooooooooooooong_table_name (event_id int NOT NULL, event_time timestamptz NOT NULL DEFAULT now());

INSERT INTO non_distributed_child_very_very_very_loooooooooooooooooooooooooooooooooong_table_name SELECT i, '2021-06-15' FROM generate_series(0,100)i;

ALTER TABLE parent ATTACH PARTITION non_distributed_child_very_very_very_loooooooooooooooooooooooooooooooooong_table_name FOR VALUES FROM ('2021-06-01') TO ('2021-07-30');
ERROR:  cannot distribute "non_distributed_child_very_very_very_looooooooooooooooooooooooo" in sequential mode because it is not empty
HINT:  If you have manually set citus.multi_shard_modify_mode to 'sequential', try with 'parallel' option. If that is not the case, try distributing local tables when they are empty.
Time: 1.581 ms
```

I'm not too concerned about this, because we are not changing the behaviour anyway. It is failing on master as well w/wout fkeys.

File: src/backend/distributed/commands/table.c (hunk 3 above)
Comment (naisila) [reply src/backend/distributed/commands/table.c:402]: > I think the major drawback of switching from Pre hook to post hook, the errors are going to be detected on the workers instead of on the parents.

I believe the only difference are the errors that may occur while distributing the partition. The other errors will be detected in the coordinator as before.

> in the following example, the error is detected on the shards where it is detected on the coordinator before the patch:

This error is detected in the coordinator right after distributing the partition. Similarly, all checks in `ATExecAttachPartition` will occur in the coordinator before sending remote "attach partition" commands for shards.

> Similarly, when we do things in the post hook, we ensure that the proper locks are acquired by Postgres on the tables. Now, they are not acquired at this step. Do you see any issues with that?

Hmm I don't see any obvious issues myself, but would it be safer if I acquired the proper locks inside the `Preprocess` function?

**Comment by naisila** [comment]
> It seems it'd be hard to follow for users in the release notes. Maybe something like Fixes a bug that prevents attaching partitions when colocated foreign key exists

Makes sense. I chose that description because, although the motivation is that bug, the fix encapsulates a lot more than that specific case. I will change the description, but leave the PR name for the developers, do you think that's ok? @onderkalaci

**Comment by naisila** [comment]
> It is annoying that even after this patch, the users could hit some issues when (a) the partition has already data (b) the name of the partition is long (hence switch to sequential execution mode). Can you see an easy way to fix that?

I guess I could use the pointers you gave me but I need to study this more and it will take some time.

> I'm not too concerned about this, because we are not changing the behaviour anyway. It is failing on master as well w/wout fkeys.

Considering this, it seems a bit out of scope for this PR. Can we open a separate issue about this one (if not already opened) and fix it later? I assume the current list of bugs I am working on has more priority than this issue on timeseries tasks? @onderkalaci

File: src/backend/distributed/commands/table.c (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:402]: > I believe the only difference are the errors that may occur while distributing the partition. The other errors will be detected in the coordinator as before.

> This error is detected in the coordinator right after distributing the partition. Similarly, all checks in > ATExecAttachPartition will occur in the coordinator when executing ALTER TABLE parent ATTACH > PARTITION child locally, which happens before sending remote "attach partition" commands for shards.


I don't think so. See below, the error is detected on the **workers** after this patch. And, all the checks in `ATExecAttachPartition ` happen similarly.
```SQL
create table part ( a int, list varchar(5) ) partition by list (list);
SELECT create_distributed_table('part', 'a');

create table part_1 ( a int, list varchar(5), x int );

ALTER TABLE part ATTACH PARTITION part_1 FOR VALUES IN (1,2);
ERROR:  table "part_1" contains column "x" not found in parent "part"
DETAIL:  The new partition may contain only the columns present in parent.
```


The order is : 1)Pre hook 2) standard process utility 3)Post hook.

Before this patch, we first call `ATExecAttachPartition` on the coordinator for the distributed tables at (2). Now, we first call `ATExecAttachPartition` on the shards at (1)


> Hmm would it be safer if I acquired the proper locks inside the Preprocess function? I believe there might be different behaviours if the parent table is concurrently changed while the partition is being distributed. Before with the post hook the parent table already has proper locks at the step of distributing the attached partition. @onderkalaci


Yes, we need to make sure that we acquire the locks in the same order as Postgres. Otherwise, we can introduce very weird deadlocks. An example for such unexpected deadlocks that occurs in Citus but not in PG: https://github.com/citusdata/citus/pull/5159

Please work with @velioglu on the locking issues and make sure that we acquire the locks in the same order as before.

**Comment by onderkalaci** [comment]
> Considering this, it seems a bit out of scope for this PR. Can we open a separate issue about this one (if not already opened) and fix it later? I assume the current list of bugs I am working on has more priority than this issue on timeseries tasks? @onderkalaci

Yes, please open an issue if not already exists and add it to #4983. Thanks

File: src/backend/distributed/commands/utility_hook.c [hunk 4]
```diff
@@ -443,6 +443,17 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
...
+	 * We only process ALTER TABLE ... ATTACH PARTITION commands in the function below
+	 * and distribute the partition if necessary.
+	 */
+	if (IsA(parsetree, AlterTableStmt))
+	{
+		AlterTableStmt *alterTableStatement = (AlterTableStmt *) parsetree;
+
+		PreprocessAlterTableStmtAttachPartition(alterTableStatement, queryString);
+	}
+
```
[the remark this answers is not in the document]
Comment (naisila) [reply src/backend/distributed/commands/utility_hook.c:456]: I believe guarding with `EnableDDLPropagation` isn't desirable, as before this fix we were not guarding it in the `PostprocessAlterTableStmtAttachPartition`.

We intentionally call `PreprocessAlterTableStmtAttachPartition` before `ops->preprocess`. In `PreprocessAlterTableStmtAttachPartition`, we only distribute the partition, and in `ops->preprocess` which calls `PreprocessAlterTableStmt`, we generate worker DDLJobs ("attach partition" commands for shards), and after local "attach partition" command has completed, we finish by executing worker DDLJobs

```
DDLJob *ddlJob = NULL;
foreach_ptr(ddlJob, ddlJobs)
{
	ExecuteDistributedDDLJob(ddlJob);
}
```

**Comment by naisila** [comment]
Locally tested similar scenarios to the one in the issue #4483 with `pg_partman`, works fine 👍

File: src/backend/distributed/commands/table.c [hunk 5]
```diff
@@ -399,19 +399,21 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
...
 	{
 		if (alterTableCommand->subtype == AT_AttachPartition)
 		{
-			Oid relationId = AlterTableLookupRelation(alterTableStatement, NoLock);
+			LOCKMODE lockmode = AlterTableGetLockLevel(alterTableStatement->cmds);
+			Oid relationId = AlterTableLookupRelation(alterTableStatement, lockmode);
 			PartitionCmd *partitionCommand = (PartitionCmd *) alterTableCommand->def;
 			bool partitionMissingOk = false;
-			Oid partitionRelationId = RangeVarGetRelid(partitionCommand->name, NoLock,
+			Oid partitionRelationId = RangeVarGetRelid(partitionCommand->name,
```
[the remark this answers is not in the document]
Comment (naisila) [reply src/backend/distributed/commands/table.c:415]: Yes Postgres uses lockLevel for the partition as well

File: src/backend/distributed/commands/table.c [hunk 6]
```diff
@@ -399,19 +399,24 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
...
+			/*
+			 * We acquire the lock on the parent and child as we are in the pre-process
+			 * and want to ensure we acquire the locks in the same order with Postgres
+			 */
+			LOCKMODE lockmode = AlterTableGetLockLevel(alterTableStatement->cmds);
+			Oid relationId = AlterTableLookupRelation(alterTableStatement, lockmode);
 			PartitionCmd *partitionCommand = (PartitionCmd *) alterTableCommand->def;
 			bool partitionMissingOk = false;
-			Oid partitionRelationId = RangeVarGetRelid(partitionCommand->name, NoLock,
+			Oid partitionRelationId = RangeVarGetRelid(partitionCommand->name, lockmode,
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:419]: I don't think that this change is strictly necessary, but there seems no downside of acquiring this lock either. We already acquire a higher lock level in `CreateDistributedTable` within the next few lines.

Still this change feels more inline with Postgres

**Review by onderkalaci — APPROVED** [review]
